glom/core.py

```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
     (e.g., ``len``, ``sum``, any ``lambda``) will not be wrapped in a
     GlomError.
     """
-    pass
+    @classmethod
+    def wrap(cls, exc):
+        class GlomWrapError(type(exc), GlomError): pass
+        return GlomWrapError(*exc.args)
+        # this is another pattern, but doesn't work for C-defined exceptions
+        # (maybe if hasattr(sys.modules[type(exc).__module__], "__file__")?)
+        exc2 = copy.copy(exc)
+        exc2.__class__ = GlomWrapError
+        return exc2
+
+    def _finalize(self, scope):
+        from . import trace
+
+        # tried MANY approaches here -- one problem pragmatically is
+        # that traceback module and sys.exc_info() see different
+        # stacks here -- maybe it is something to do with running
+        # under pytest
+        lines = traceback.format_exc().split("\n")
+        limit = 0
+        for line in reversed(lines):
+            if __file__ in line and "_glom" in line:
+                limit -= 1
+                break
+            limit += 1
+
+        self.short_stack = trace.short_stack(scope)
+        self.inner_format = ("\n".join(
+            ["", trace.short_stack(scope)] + lines[-limit:]))
+
+    def __str__(self):
+        if hasattr(self, "inner_format"):
+            return self.inner_format
+        return ""
+
+    def __repr__(self):
+        return str(self)
```

Comment:
Why have `__repr__` be the same as `__str__`? Exceptions often have these differ (notably, `KeyError`, for better or worse).

Reply:
I have no idea -- I was just trying to get the stack trace and repr to look right so I could do testing :-)